rfcs/027-pipeline-intermediate-storage/README.md

```diff
@@ -0,0 +1,147 @@
+
+## Requirements for Intermediate Store
+
+Taking into account the business logic of our services, and given the issues with the current VHS highlighted in [Problem](#problem), there are a number of requirements and desirable features for any alternative store:
+
+1. The ability to insert items into the store with some unique string ID and some specific integer version (set externally to the pipeline from our adapters), with a method to only store data when the version given is greater or equal to what is currently in the store.
+
+2. The ability to retrieve data from the store, where the latest data is always retrieved and the version is returned with the data.
+
+3. Happily stores data in our current range of around 8GB, and scaling beyond this for future proofing.
+
+4. Some form of disk persistence so that we do not need to perform a reindex from scratch in the case of the database going down.
+
+5. Cheaper overall cost than currently, including read / write requests, associated networks costs, and storage costs.
+
+6. Ideally improved performance.
+
+7. Ideally better visibility into the data.
+
+## Store Options
+
+There are a number of different options to use as an intermediate store that I have considered, outlined below.
+
+### Elasticsearch
+
+Pros:
+
+* Used elsewhere in the pipeline so general familiarity with it.
+
+* Kibana would provide a level of visibility into the output at different stages unlike anything else available, and the ability to perform complex queries on our data would likely be an asset to debugging.
+
+* Can access via AWS network address to reduce NAT gateway costs (see [here](https://www.elastic.co/guide/en/cloud/current/ec-getting-started-private.html)).
+
+* Back of the envelope calculations show likely to be much cheaper than current solution.
+
+* Horizontal scalability means quite capable of supporting data at sizes much beyond what we have now.
+
+Cons:
+
+* Requires use of external cloud platform or maintenance of our own Elasticsearch cluster, rather than being a service within AWS.
+
+* Terraform does not control the creation and removal of Elasticsearch indices so might require manual clean up after removing a pipeline.
+
+* No transactions which might prove useful for certain use cases.
+
+Questions:
+
+* Elasticsearch supports [versioning](https://www.elastic.co/blog/elasticsearch-versioning-support), including the ability to index with our own external version numbers. However when using external versioning it only indexes data when the version number is greater than currently stored, whereas our business logic currently expects updates to be processed when versions are greater or equal. Some thought will need to be given whether this might work for us, and if not we would need to implement our own system on top of the Elasticsearch APIs (which may not be particularly resilient to race conditions due to lack of transactions).
```

Comment:
You can use [`external_gte`](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-index_.html#index-version-types), which we use in a few places.

Reply:
Oooo exactly what I was looking for but didn't find.

Reply:
I think we already use external_gte in the ingestor

Reply:
> I think we already use external_gte in the ingestor

[Yarp](https://github.com/wellcomecollection/catalogue/blob/master/pipeline/ingestor/ingestor_common/src/main/scala/uk/ac/wellcome/platform/ingestor/common/Indexer.scala#L28)